### PR TITLE
Introduce prelink and postlink commands

### PR DESCRIPTION
**Pull request by Kureev**
As we discussed in Discord, I've made a standalone repo for hooks wrapper.
Basically, since now, users can make their own plugins with a hooks support.

## How it works
Basically, I moved `makeHook` function to the standalone repo to allow other people use it.

## Scope of this PR
Replace in-place function by the `rnpm-hooks` module. Added missing `require` for testing and adjusted link code to take care about generating hooks. I'd like to keep `rnpm-hooks` as tiny as possible (and not to depend on our `dependency` object structure), so I think [this kind of logic](https://github.com/rnpm/rnpm-plugin-link/compare/chore/better-hooks?expand=1#diff-1dbe17fda47e0a0d65b16ad4646552d0R103) should be scoped in this specific plugin.

See [rnpm-hooks](https://github.com/rnpm/rnpm-hooks) for the further reading.

### CODE REVIEW

**Comment by grabbou** [comment]
Yeah, that works :)

Have you thought about moving async to hooks repo as well so that we just:
```js
const hooks = require('hooks')();
hooks.before([...])
hooks.after([...])
hooks.run(link, cb)
```
?

**Comment by Kureev** [comment]
Can you explain how it'll benefit for us? It looks like an `async` interface analog but + state machine (why?). Or you have any cases where this hooks s/m may be handy for us?

**Comment by grabbou** [comment]
Then I guess the name shouldn't be hooks but something like `func-utils` or maybe just `utils`. The whole point was to avoid writing async all over in every project that wants to use hooks, but do:
```js
hooks.run([before, command, after])`
```
instead (the previous example was quite shit, ignore it)

File: src/link.js [hunk 1]
```diff
@@ -113,11 +95,29 @@ module.exports = function link(config, args, callback) {
...
-    const link = makeLink(dependency);
+    const queue = [];
+
+    var prelink;
+    var postlink;
+
+    if (dependency.config.hooks) {
+      prelink = dependency.config.hooks.prelink;
+      postlink = dependency.config.hooks.postlink;
+    }
```
Comment (grabbou) [inline src/link.js]: Can't we just
```
const prelink = dependency.config.hooks.prelink;
const postlink = dependency.config.hooks.postlink;
```
?

File: src/link.js [hunk 2]
```diff
@@ -113,11 +95,29 @@ module.exports = function link(config, args, callback) {
...
+    queue.push(makeLink(dependency));
+
+    if (postlink) {
+      queue.push(makeHook(postlink));
+    }
+
+    queue.push(next);
 
-    async.waterfall([prelink, link, postlink, next]);
+    async.waterfall(queue);
```
Comment (grabbou) [inline src/link.js]: There's no `callback` to waterfall. I think it should be `async.waterfall([...], next)` instead.

File: src/link.js [hunk 3]
```diff
@@ -113,11 +95,29 @@ module.exports = function link(config, args, callback) {
...
+    var prelink;
+    var postlink;
+
+    if (dependency.config.hooks) {
+      prelink = dependency.config.hooks.prelink;
+      postlink = dependency.config.hooks.postlink;
+    }
+
+    if (prelink) {
+      queue.push(makeHook(prelink));
```
Comment (grabbou) [inline src/link.js]: as I said in the comments, maybe something with `command` name would've been better

File: src/link.js (hunk 1 above)
Comment (Kureev) [inline src/link.js]: if we don't have any hooks, the code above will fail (Can't get prelink of undefined). That's why I wrapped it into check.

File: src/link.js (hunk 3 above)
Comment (Kureev) [inline src/link.js]: I totally agree with it. That was a point I wanted to discuss with you yesterday when you went offline :+1:

File: src/link.js (hunk 2 above)
Comment (Kureev) [inline src/link.js]: Oh shit, true!

**Comment by Kureev** [comment]
so something like this:
```js
const makeCommand = require('rnpm-commands').makeCommand;
const run = require('rnpm-commands').run;

const someCommand = makeCommand('some-shell-command-here');
const pluginFunction = () => {...};

run([someCommand, pluginFunction]);
```

But in this case we still need to have an `async.series` to run it. So I'd prefer to stick to one `makeCommand` for now. If people will request a `run` command or something like it - we can add it at any time (but I don't think it'll value to them).

**Comment by Kureev** [comment]
> Maybe it could have two functions - wrapShellCommand
> ```js
> const before = beforeLink.map(wrapShellCommand);
> ```
> to return an array of functions with spawn we can execute and - run

Can you explain how we may have a list of commands here? One string always maps to one command. There's no point to have an array here. Do I miss something?

**Comment by Kureev** [comment]
> I guess that once demand is higher, map(wrapShellCommand) can be moved here 
> https://github.com/rnpm/rnpm/blob/master/src/config/index.js#L22 and
> https://github.com/rnpm/rnpm/blob/master/src/config/index.js#L38
> so you can just run them w/o thinking about wrapping them.

So you want to wrap them automatically? Well, sounds good. Indeed, we can wrap all custom commands even before we run a plugin :+1: I'll take care about it!

**Comment by grabbou** [comment]
> Can you explain how we may have a list of commands here? One string always maps to one command. There's no point to have an array here. Do I miss something?

My bad, for some reason I assumed it's an array :dancer:

**Comment by Kureev** [comment]
@grabbou can you review it? :8ball:

File: src/link.js [hunk 4]
```diff
@@ -113,11 +93,27 @@ module.exports = function link(config, args, callback) {
...
-    const link = makeLink(dependency);
+    const queue = [];
+
+    var prelink;
+    var postlink;
+
+    if (dependency.config.commands) {
+      prelink = dependency.config.commands.prelink;
+      postlink = dependency.config.commands.postlink;
+    }
```
Comment (grabbou) [inline src/link.js]: What happens when these are not defined? Is `asycn.waterfall` capable of handling undefined `prelink`, `postlink`? I think commands are **always** defined so we should check for presence of prelink/postlink instead

Comment (Kureev) [inline src/link.js]: yes, if it's undefined - it'll skip it.

Comment (grabbou) [inline src/link.js]: My suggestion for these lines is just:
```js
async.series(
   dependencies.map(dependency => next =>
      async.waterfall([
         dependency.commands.preLink,
         makeLink(...),
         dependency.commands.postLink
     ], next)
   ), 
   callback
);
```

**Comment by Kureev** [comment]
I. JUST. CANT. BELIVE!

**Comment by grabbou** [comment]
Wait for your first contract and you will know that pain haha!

**Comment by Kureev** [comment]
Released `1.4.0`

**Pull request merged**